### PR TITLE
Basic framework for DATEDIFF

### PR DESCRIPTION
**Pull request by norton120**
## WHY?
we have `DATEDIFF` all over our code, it is not x database

## WHAT CHANGES?
This adds the basic framework from norton120/feature/DATEDIFF/a that will carry the rest of the work

### CODE REVIEW

File: macros/datediff.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+         - left_val (date/timestamp) the value before the minus in the equation "left - right"
+         - right_val (date/timestamp) the value after the minus in the equation "left - right"
+         - date_format (pattern) a string pattern for the provided arguments (primarily for BigQuery)
+       RETURNS: An integer representing the delta in `part` units
+    #}
+
+    {%if part not in ('day','week','month','year',) %}
+        {{exceptions.raise_compiler_error("macro datediff for target does not support date part value " ~ part)}}
+    {%- endif -%}
+
```
Comment (Dylan-Health-Union) [inline macros/datediff.sql]: Worth adding a test to see if coercing to date fails and throw an error?

File: macros/datediff.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {%- if target.type ==  'postgres' -%} 
+        {%- if part == 'year' -%}
```
Comment (Dylan-Health-Union) [inline macros/datediff.sql]: any need for a part|lower here?

File: macros/datediff.sql [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    {%- elif target.type == 'bigquery' -%}
+
+        DATE_DIFF(PARSE_DATE('{{date_format}}', {{left_val}}), 
+                  PARSE_DATE('{{date_format}}', {{right_val}}), 
+                  {{part|upper}})
+
+    {%- elif target.type == 'snowflake' -%}
+
+        DATEDIFF('{{part|lower}}', {{right_val}},{{left_val}})
```
Comment (Dylan-Health-Union) [inline macros/datediff.sql]: If we're coercing to date / timestamp in pg, any value to doing it for bq / snowflake? Also, why not coerce them once (I'm assuming both also support timestamp / dates) to remove the need to constantly cast?

File: macros/datediff.sql (hunk 1 above)
Comment (norton120) [reply macros/datediff.sql]: Yep and actually bad values too. Good call

Comment (norton120) [reply macros/datediff.sql]: I don't know that DBT has a clean way to test for exceptions? if they do it's new. gotta dig a little, I'll add that in a later PR

File: macros/datediff.sql (hunk 3 above)
Comment (norton120) [reply macros/datediff.sql]: so they don't all behave the same with coercing into timestamp / date, unfortnately :( the same TS values coerced the same between Snowflake and PG will give different diff values, which is gross. Bigquery does it's own weirdness as you can see with parsing first from strings. it behaves more like raw JS. 

however I think there is room to standardize here for sure. I will tackle in a new PR

File: macros/datediff.sql (hunk 1 above)
Comment (norton120) [reply macros/datediff.sql]: Update - there's no effective method to test _postitive_ for exceptions at runtime in DBT :( that's a shame, but I'm not seeing an easy way to test these :( :( . We could build a runtime pytest suite (wouldn't take much) but that's some scope ➕ ➕  for sure.

File: macros/datediff.sql (hunk 2 above)
Comment (norton120) [reply macros/datediff.sql]: per your comment above I updated it to lower the parts arg across the board - so it's nice and dry now :the_horns:

File: macros/datediff.sql (hunk 3 above)
Comment (norton120) [reply macros/datediff.sql]: Lemme try to flatten this in the next PR when I add the h/m/s datediff. it'll be more obvious where we can dry it up in there I think.

**Pull request merged**